units/pod_definition_cleanup.go

```diff
@@ -0,0 +1,139 @@
+package units
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/evergreen-ci/cocoa"
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/cloud"
+	"github.com/evergreen-ci/evergreen/model/pod/definition"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+const (
+	podDefinitionCleanupJobName = "pod-definition-cleanup"
+	podDefinitionTTL            = 30 * 24 * time.Hour
+)
+
+func init() {
+	registry.AddJobType(podDefinitionCleanupJobName, func() amboy.Job {
+		return makePodDefinitionCleanupJob()
+	})
+}
+
+type podDefinitionCleanupJob struct {
+	job.Base `bson:"metadata" json:"metadata" yaml:"metadata"`
+
+	env       evergreen.Environment
+	settings  evergreen.Settings
+	ecsClient cocoa.ECSClient
+	podDefMgr cocoa.ECSPodDefinitionManager
+}
+
+func makePodDefinitionCleanupJob() *podDefinitionCleanupJob {
+	j := &podDefinitionCleanupJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    podDefinitionCleanupJobName,
+				Version: 0,
+			},
+		},
+	}
+	return j
+}
+
+// NewPodDefinitionCleanupJob creates a job that cleans up pod definitions that
+// have not been used recently.
+func NewPodDefinitionCleanupJob(id string) amboy.Job {
+	j := makePodDefinitionCleanupJob()
+	j.SetID(fmt.Sprintf("%s.%s", podDefinitionCleanupJobName, id))
+	return j
+}
+
+func (j *podDefinitionCleanupJob) Run(ctx context.Context) {
+	defer func() {
+		j.MarkComplete()
+
+		if j.ecsClient != nil {
+			j.AddError(errors.Wrap(j.ecsClient.Close(ctx), "closing ECS client"))
+		}
+	}()
+	if err := j.populate(); err != nil {
+		j.AddError(err)
+		return
+	}
+
+	podDefs, err := definition.FindByLastAccessedBefore(podDefinitionTTL, -1)
+	if err != nil {
+		j.AddError(err)
+		return
+	}
+
+	for _, podDef := range podDefs {
+		if podDef.ExternalID == "" {
+			if err := podDef.Remove(); err != nil {
```

Comment:
I may be missing something - wouldn't we want to remove all pod definitions that have passed TTL?

Reply:
We do. This loop removes all the pod definitions that pass their TTL regardless of whether they're tied to an external ID or not. The first if on L80 handles a future-proofing case where a pod definition exists but is not associated with an external pod definition. That can't supposedly happen now, but I want to make sure that if a pod definition like that could exist, it gets deleted by the cleanup job as well since it serves no purpose.

Reply:
It looks like we're only removing the pod definitions from our own collection in this future proofing case - shouldn't we call `Remove` no matter what then?

Reply:
Since the pod definition manager accepts an external storage cache, it's implicitly deleting the DB doc [as part of the manager functionality](https://github.com/evergreen-ci/cocoa/blob/df9129525b0266148e1191aa3dd2a1d410fb208f/ecs/pod_definition_manager.go#L159-L161), so [this](https://github.com/kimchelly/evergreen/blob/c2a6d2600c5146ea1254d35cae5281152c825c8a/model/pod/definition/definition.go#L85-L95) is called by the pod definition manager. I tested that this is working correctly [here](https://github.com/Kimchelly/evergreen/blob/c2a6d2600c5146ea1254d35cae5281152c825c8a/units/pod_definition_cleanup_test.go#L46-L48).

Reply:
Gotcha, my bad

Reply:
No problem, I realize that it might not be that clear just from a glance.